crates/dyn-abi/src/coerce.rs

```diff
@@ -0,0 +1,861 @@
+            s.parse::<U256>()
+                .map_err(|e| ErrMode::from_external_error(input, ErrorKind::Verify, e))?
+                .checked_mul(U256::from(10usize.pow(units as u32)))
+                .ok_or_else(|| {
+                    ErrMode::from_external_error(input, ErrorKind::Verify, Error::IntOverflow)
+                })
+        }?;
+
+        if uint.bit_len() > len {
+            return Err(ErrMode::from_external_error(
+                input,
+                ErrorKind::Verify,
+                Error::IntOverflow,
+            ))
+        }
+
+        Ok(uint)
+    })
+}
+
+#[inline]
+fn uint_integer(input: &mut &str) -> PResult<()> {
+    let has_prefix = matches!(
+        input.get(..2),
+        Some("0b" | "0B" | "0o" | "0O" | "0x" | "0X")
+    );
+    if has_prefix {
+        *input = &input[2..];
+        hex_digit1(input)
+    } else {
+        digit1(input)
+    }
+    .map(drop)
+}
+
+#[inline]
+fn uint_units(input: &mut &str) -> PResult<usize> {
+    alpha0(input).and_then(|s| {
+        Ok(match s {
+            "ether" => 18,
+            "gwei" | "nano" | "nanoether" => 9,
+            "" | "wei" => 0,
+            _ => return Err(ErrMode::from_error_kind(input, ErrorKind::Fail)),
+        })
+    })
+}
+
+#[inline]
+fn fixed_bytes<'i>(len: usize) -> impl Parser<&'i str, Word, ContextError> {
+    #[cfg(feature = "debug")]
+    let name = format!("bytes{len}");
+    #[cfg(not(feature = "debug"))]
+    let name = "bytesN";
+    trace(name, move |input: &mut &str| {
+        if len > Word::len_bytes() {
+            return Err(ErrMode::from_external_error(
+                input,
+                ErrorKind::Fail,
+                Error::InvalidFixedBytesLength(len),
+            ))
+        }
+
+        let hex = hex_str(input)?;
+        let mut out = Word::ZERO;
+        match hex::decode_to_slice(hex, &mut out[..len]) {
+            Ok(()) => Ok(out),
+            Err(e) => Err(hex_error(input, e)),
+        }
+    })
+}
+
+#[inline]
+fn address(input: &mut &str) -> PResult<Address> {
+    trace("address", fixed_bytes_inner)
+        .parse_next(input)
+        .map(Address::from)
+}
+
+#[inline]
+fn function(input: &mut &str) -> PResult<Function> {
+    trace("function", fixed_bytes_inner)
+        .parse_next(input)
+        .map(Function::from)
+}
+
+#[inline]
+fn bytes(input: &mut &str) -> PResult<Vec<u8>> {
+    trace("bytes", hex_str)
+        .parse_next(input)
+        .map(|s| hex::decode(s).unwrap())
+}
+
+#[inline]
+fn string(input: &mut &str) -> PResult<String> {
+    trace("string", string_inner)
+        .parse_next(input)
+        .map(String::from)
+}
+
+#[inline]
+fn string_inner<'i>(input: &mut &'i str) -> PResult<&'i str> {
+    let Some(delim) = input.chars().next() else {
+        return Ok("")
+    };
+    let has_delim = matches!(delim, '"' | '\'');
+    if has_delim {
+        *input = &input[1..];
+    }
+
+    // TODO: escapes?
+    let min = if has_delim { 0 } else { 1 };
```

Comment:
hmm, good question, maybe this could happen with same escaped json inputs, but I think we can ignore that